### ISSUE TITLE
[dv/otbn] chip level scramble test failure

### ISSUE THREAD

**Issue by cindychip** [lowRISC/opentitan#13974]
Hi Douglas,

I am looking at the `X` failure in otbn_scramble test, and have two questions to confirm with you:

1). Will `imem_sec_wipe` wipe out the actual i_mem to some random value or just the keys and some associated CSRs?

2). This is related to where the `X` generated:
     From the c test, looks like the test did the following things
     a). Randomly pick some imem and dmem to write.
     b). Read back the written imem and dmem and check read results.
     c). Re-fetch otp scrambling keys.
     d). Re-read the imem and dmem address from step a).
     If my understanding is correct - since step c) re-scrambles, step d) is very unlikely to pick the written addresses. Then without initializing the imem/dmem values, we are very likely to read out some un-initialized memory and get `X`.
     Do we miss some imem/dmem initialization steps? Or should mem_sec_wipe re-assign the entire memory?

ERROR MESSAGE:
```
## Failure Buckets

* `Offending '(((((((((!d_mask[*]) || (d_mask[*] && (!$isunknown(d_data[(* * *)+:*])))) && ((!d_mask[*]) || (d_mask[*] && (!$isunknown(d_data[(* * *)+:*]))))) && ((!d_mask[*])
 || (d_mask[*] && (!$isunknown(d_data[(* * *)+:*]))))) && ((!d_mask[*]) || (d_mask[*] && (!$isunknown(d_data[(* * *)+:*]))))) && ((!d_mask[*]) || (d_mask[*] && (!$isunknown(d_
data[(* * *)+:*]))))) && ((!d_mask[*]) || (d_mask[*] && (!$isunknown(d_data[(* * *)+:*]))))) && ((!d_mask[*]) || (d_mask[*] && (!$isunknown(d_data[(* * *)+:*]))))) && ((!d_mas
k[*]) || (d_mask[*] && (!$isunknown(d_data[(* * *)+:*])))))'` has 1 failures:
    * Test chip_sw_otbn_mem_scramble has 1 failures.
        * 0.chip_sw_otbn_mem_scramble.2409556210\
          Line 582, in log /usr/local/google/home/chencindy/opentitan/scratch/master/chip_earlgrey_asic-sim-vcs/0.chip_sw_otbn_mem_scramble/out/run.log

                	Offending '(((((((((!d_mask[0]) || (d_mask[0] && (!$isunknown(d_data[(8 * 0)+:8])))) && ((!d_mask[1]) || (d_mask[1] && (!$isunknown(d_data[(8 * 1)+:8])
)))) && ((!d_mask[2]) || (d_mask[2] && (!$isunknown(d_data[(8 * 2)+:8]))))) && ((!d_mask[3]) || (d_mask[3] && (!$isunknown(d_data[(8 * 3)+:8]))))) && ((!d_mask[4]) || (d_mask[
4] && (!$isunknown(d_data[(8 * 4)+:8]))))) && ((!d_mask[5]) || (d_mask[5] && (!$isunknown(d_data[(8 * 5)+:8]))))) && ((!d_mask[6]) || (d_mask[6] && (!$isunknown(d_data[(8 * 6)
+:8]))))) && ((!d_mask[7]) || (d_mask[7] && (!$isunknown(d_data[(8 * 7)+:8])))))'
                UVM_ERROR @ 4864.601236 us: (tlul_assert.sv:257) [ASSERT FAILED] dDataKnown_M
                UVM_INFO @ 4864.601236 us: (uvm_report_catcher.svh:705) [UVM/REPORT/CATCHER]
                --- UVM Report catcher Summary ---
```

Thanks,
Cindy

**Comment by engdoreis** [comment]
Hi Cindy,

Your analysis about the test design is accurate, but I would add an additional step.
   e) Expect an ECC error by checking whether an exception has occurred.

This was the original design of the tests, however Ibex/OTBN changed to trigger an internal interrupt in case of ECC error rather than an exception so I created the #13954.
But even though the test is passing just with Verilator (Neither VCS nor FPGA are passing). And now I know the reason due to your explanation
I believe that the behavior seen on Verilator should be replicated for vcs and FPGA, I mean, if the SW performs a mem_sec_wipe it should get an internal irq if it tries to read the memory.

Please Let me know if that answer your question
Thanks,

**Comment by cindychip** [comment]
Thank you so much for the explanation. And thanks for pointing to the fix.
Now I understand - you are expecting to see an interrupt when reading the re-scrambled memory, but in DV, the assertion error hits first because you are reading out `X` (uninitialized value)...

Just curious how you plan to resolve this issue?
Do you plan to initialize all the imem and dmem in step a), or find out a way to disable this assertion?

**Comment by engdoreis** [comment]
@GregAC Could you let me know your thoughts on a possible approach?

**Comment by vogelpi** [comment]
Hi,
I had a look at this as part of debugging #14196 . As @cindychip correctly pointed out, due to the re-scrambling (which also changes the addressing order of the memories) it's highly unlikely to actually read initialized addresses in step d) .

Initializing the entire DMEM and IMEM sounds reasonable but for this to work we would also have to load and run an OTBN application that initializes the scratchpad part of DMEM. Because this part of DMEM is not accessible to Ibex. But due to the re-scrambling Ibex might suddenly get access to this part and again read X.

Finally, as @engdoreis suggested, OTBN will produce an integrity error upon reading the first IMEM/DMEM address anyway due to the ECC error. So maybe it's fine to adapt the test as follows:
1. Initialize the entire IMEM.
2. When reading out IMEM after re-scrambling expect an integrity error and be done with the test. I.e., don't try to read DMEM at all and thus also don't worry about initializing the entire DMEM.

I'll give this a try.

**Comment by GregAC** [comment]
> don't try to read DMEM at all and thus also don't worry about initializing the entire DMEM.

I think we should try to test both the IMEM and DMEM scrambling. We're happy the primitive is fine what this is testing is the integration and there we could have messed up the IMEM but not the DMEM and vice versa.

The suggestion of initialising the DMEM using a simple OTBN programs sounds reasonable to me. The following should do the trick

```
.text

wipe_dmem:
  li x2, 0
  loopi 128, 1
    bn.sid x0, 0(x2++)
  ecall
```

It will fill the dmem with whatever happens to be in register w0.

**Comment by vogelpi** [comment]
Thanks @GregAC for your input. I'll try this out. After including @engdoreis ' fix from #13954 and zeroing the entire IMEM and the bus accessible DMEM part, the test fails with an integrity failure:
```
UVM_ERROR @ 9577.005614 us: (cip_tl_seq_item.sv:223) [req] d_user.data_intg act (0x0) != exp (0x79)
UVM_INFO @ 9577.005614 us: (uvm_report_catcher.svh:705) [UVM/REPORT/CATCHER]
```

**Comment by engdoreis** [comment]
Thanks @GregAC and @vogelpi, I started with `IMEM` to see whether the suggested approach would work, and there's been some progress.
When I run the test with the memory initialized I get a different error. It seems to me that some `sv` code is asserting that the value doesn't match before the internal irq fires.
```
* `UVM_ERROR @ * us: (cip_tl_seq_item.sv:223) [req] d_user.data_intg act (*) != exp (*)` has 1 failures:
    * Test chip_sw_otbn_mem_scramble has 1 failures.
        * 0.chip_sw_otbn_mem_scramble.3082500892\
          Line 1252, in log /home/doreis/lowrisc/git/opentitan/scratch/fix_otbn_scramble_chip_test/chip_earlgrey_asic-sim-vcs/0.chip_sw_otbn_mem_scramble/out/run.log

                UVM_ERROR @ 8335.077958 us: (cip_tl_seq_item.sv:223) [req] d_user.data_intg act (0x64) != exp (0x54)
                UVM_INFO @ 8335.077958 us: (uvm_report_catcher.svh:705) [UVM/REPORT/CATCHER]
                --- UVM Report catcher Summary ---
```
What do you suggest?

**Comment by tjaychen** [comment]
@engdoreis how are you doing the memory initialization? the `otbn` memory is stored with integrity that is directly forwarded. Are you backdoor initializing it? or going through the CPU to load all 0's?

**Comment by engdoreis** [comment]
@tjaychen I doing by SW using the DIF.
```
 uint32_t value = 0xA5A5A5A5;
  for(uint32_t offset = 0; offset < OTBN_IMEM_SIZE_BYTES; offset += sizeof(value)){
    CHECK_DIF_OK(dif_otbn_imem_write(&ctx->dif, offset, &value, sizeof(value)));
  }
```

**Comment by tjaychen** [comment]
i'm sorry, i think i completely misunderstood the point of this issue (skipped over a few comments in the beginning).
so sounds like what we're saying is that we actually expect integrity failures, since the initialization done in the beginning no longer applies after requesting for a new scramble key correct?

If that's the case I think we just need to update the DV side to either disable that check or expect an integrity error.  @cindychip is that right?

**Comment by cindychip** [comment]
> i'm sorry, i think i completely misunderstood the point of this issue (skipped over a few comments in the beginning). so sounds like what we're saying is that we actually expect integrity failures, since the initialization done in the beginning no longer applies after requesting for a new scramble key correct?
> 
> If that's the case I think we just need to update the DV side to either disable that check or expect an integrity error. @cindychip is that right?

Hi Tim,

1). At the beginning, the issue is that we did not initialize the memory, but relies on write and read back randomly address. Then when we re-scramble data and address to read again, we are likely to choose an uninitialized address.

2). Now I believe (if I understand this part correctly) we moved forward to randomize the entire IMEM, the issue comes at DV interface and get the UVM_ERROR. 

@engdoreis for this error below, since it is expected, I think we can disable the DV sequence check! Let me check with the team to see if we have any existing flags to disable it.

``
* `UVM_ERROR @ * us: (cip_tl_seq_item.sv:223) [req] d_user.data_intg act (*) != exp (*)` has 1 failures:
    * Test chip_sw_otbn_mem_scramble has 1 failures.
        * 0.chip_sw_otbn_mem_scramble.3082500892\
          Line 1252, in log /home/doreis/lowrisc/git/opentitan/scratch/fix_otbn_scramble_chip_test/chip_earlgrey_asic-sim-vcs/0.chip_sw_otbn_mem_scramble/out/run.log
                UVM_ERROR @ 8335.077958 us: (cip_tl_seq_item.sv:223) [req] d_user.data_intg act (0x64) != exp (0x54)
                UVM_INFO @ 8335.077958 us: (uvm_report_catcher.svh:705) [UVM/REPORT/CATCHER]
                --- UVM Report catcher Summary ---
```

**Comment by cindychip** [comment]
@engdoreis, @weicaiyang suggested to try adding this to chip_sim_cfg.hjson your test entry:
`run_opts: ["+en_scb_tl_err_chk=0"]`
This should disable scb check for integrity error.
Example:
https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/dv/chip_sim_cfg.hjson#L530

**Comment by engdoreis** [comment]
Thanks @cindychip, it worked.

However there's another issue, dvsim is failing due to timeout.
```
          [   legend    ]: [Q: queued, D: dispatched, P: passed, F: failed, K: killed, T: total]                       
00:03:36  [    build    ]: [Q: 0, D: 0, P: 1, F: 0, K: 0, T: 1] 100%                                                   
01:03:36  [     run     ]: [Q: 0, D: 0, P: 0, F: 0, K: 1, T: 1] 100% 
```
As a reference, this is the verilator time:
```
real    4m3.980s
user    15m59.019s
sys     0m13.686s
```
But what I don't understand is that the dvsim log shows that vcs simulated 6911.794856 us, which should take no more than 10 minutes
```
tb.dut.u_ast.u_io_clk.u_io_osc: IO Clock Power-up Frequency: 25887956 Hz
UVM_INFO @   2.204015 us: (mem_bkdr_util.sv:522) [mem_bkdr_util[FlashBank0Data]] Set memory
UVM_INFO @   2.204015 us: (mem_bkdr_util.sv:522) [mem_bkdr_util[FlashBank1Data]] Set memory
UVM_INFO @   2.204015 us: (mem_bkdr_util.sv:531) [mem_bkdr_util[RamRet0]] Randomizing mem contents
UVM_INFO @   2.204015 us: (tb.sv:507) [mem_bkdr_util[Rom]] Loading mem from file:
test_rom_sim_dv.39.scr.vmem
UVM_INFO @   2.204015 us: (tb.sv:448) [mem_bkdr_util[FlashBank0Data]] Loading mem from file:
otbn_mem_scramble_test_prog_sim_dv.64.scr.vmem
UVM_INFO @ 2842.884856 us: (sw_test_status_if.sv:51) [tb.u_sim_sram.u_sim_sram_if.u_sw_test_status_if] SW test transitioned to SwTestStatusInBootRom.
UVM_INFO @ 3341.174856 us: (sw_logger_if.sv:365) [tb.u_sim_sram.u_sim_sram_if.u_sw_logger_if.construct_log_and_print.unnamed$$_0.unnamed$$_1.unnamed$$_2.isolation_thread.unnamed$$_0.unnamed$$_1] String arg at addr ff60: ff60
UVM_INFO @ 3341.174856 us: (sw_logger_if.sv:519) [test_rom_sim_dv(sw/device/lib/testing/test_rom/test_rom.c:88)] ff60
UVM_INFO @ 3345.644856 us: (sw_logger_if.sv:519) [test_rom_sim_dv(sw/device/lib/testing/test_rom/test_rom.c:137)] Test ROM complete, jumping to flash!
UVM_INFO @ 3356.254856 us: (sw_test_status_if.sv:51) [tb.u_sim_sram.u_sim_sram_if.u_sw_test_status_if] SW test transitioned to SwTestStatusInTest.
UVM_INFO @ 6911.474856 us: (sw_logger_if.sv:519) [otbn_mem_scramble_test_prog_sim_dv(w/device/lib/testing/test_framework/status.c:28)] PASS!
UVM_INFO @ 6911.794856 us: (sw_test_status_if.sv:51) [tb.u_sim_sram.u_sim_sram_if.u_sw_test_status_if] SW test transitioned to SwTestStatusPassed.
UVM_INFO @ 6911.794856 us: (sw_test_status_if.sv:72) [tb.u_sim_sram.u_sim_sram_if.u_sw_test_status_if] ==== SW TEST PASSED ====
UVM_INFO @ 6911.794856 us: (chip_sw_base_vseq.sv:214) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.chip_sw_base_vseq] SW TEST PASSED!
```

**Comment by tjaychen** [comment]
does this expected error generate an alert in the test? If yes, you either need to handle the alert, or tell the DV environment to not gate on that check as well. 
I think this is the option `run_opts: ["+bypass_alert_ready_to_end_check=1"]`

**Comment by vogelpi** [comment]
Although instructing Ibex and OTBN to write IMEM and DMEM works, I think it would actually be better initialize these memories in VCS/Xcelium only. This will be a lot faster in simulation. I've now filed a PR that allows doing this selectively here: #14355 .

**Comment by engdoreis** [comment]
That makes much more sense to me, as it is only a requirement for dvsim (because of the X), and as a consequence we would save some time in the Verilator test.

**Comment by vogelpi** [comment]
I've now merged the PR to initialize IMEM and DMEM to all zero. In addition, I've rebased #14355 and included all uncontroversial changes there. The test still fails and it's very hard to debug, most likely because there is always a chance that after changing the scrambling key there is a certain chance that the ECC bits are correct and no IRQ is triggered. The test fails if a single read at the end doesn't generate an IRQ. The randomization/scrambling keys differs between VCS, Verilator, FPGA...

Thus I suggest to rework the test after merging #14355 as follows:
1. Read a couple of (say 10) random values from RV_CORE_IBEX.RND_DATA (there is a DIF for that) and from that derive offsets within IMEM and bus-accessible DMEM.
2. Wait for OTBN to be idle.
3. Write more random numbers (again taken from RV_CORE_IBEX.RND_DATA) to these offsets in IMEM and DMEM.
4. Read the numbers back and check them. All numbers must match.
5. Perform the secure wipe.
6. Wait for OTBN to be idle.
7. Read the numbers back and check them. Most reads should trigger an IRQ, but some may not as there is always a chance for the ECC bits to be correct despite the secure wipe. So the checking routine must allow some reads (say 1 out of 10)  to not trigger an ECC integrity error/interrupt.
